File: EXAMEN_BANQUITO_REST_JAVA/02. CLICON/CLICON_Comercializadora_RESTJAVA/src/main/java/ec/edu/pinza/clicon/controllers/AuthController.java
package ec.edu.pinza.clicon.controllers;

import ec.edu.pinza.clicon.models.LoginResponseDTO;
import ec.edu.pinza.clicon.models.UsuarioSesion;
import ec.edu.pinza.clicon.services.ComercializadoraRestClient;

/**
 * Controlador de autenticación dinámico via REST API.
 * ADMIN: MONSTER / MONSTER9 (puede vender a cualquier cédula)
 * CLIENTES: cédula / abcd1234 (solo pueden comprar para su propia cédula)
 */
public class AuthController {

    private final ComercializadoraRestClient restClient;

    public AuthController(ComercializadoraRestClient restClient) {
        this.restClient = restClient;
    }

    public LoginResponseDTO login(String usuario, String password) {
        return restClient.login(usuario, password);
    }
    
    public UsuarioSesion crearSesion(LoginResponseDTO response) {
        if (response != null && response.isExitoso()) {
            return new UsuarioSesion(
                response.getUsername(),
                response.getRol(),
                response.getCedula()
            );
        }
        return null;
    }
}

File: EXAMEN_BANQUITO_REST_JAVA/02.CLIMOV/app/src/main/java/com/example/climov_comercializadora_restjava/controllers/AuthController.kt
package com.example.climov_comercializadora_restjava.controllers

import com.example.climov_comercializadora_restjava.models.LoginResponseDTO
import com.example.climov_comercializadora_restjava.services.ComercializadoraApi
import com.google.gson.Gson

/**
 * Controlador de autenticación dinámico via REST API.
 * ADMIN: MONSTER / MONSTER9 (puede vender a cualquier cédula)
 * CLIENTES: cédula / abcd1234 (solo pueden comprar para su propia cédula)
 */
class AuthController(private val api: ComercializadoraApi) {

    suspend fun login(usuario: String, password: String): LoginResponseDTO {
        return try {
            val request = mapOf("username" to usuario, "password" to password)
            val response = api.login(request)
            
            if (response.isSuccessful) {
                response.body() ?: LoginResponseDTO(
                    exitoso = false,
                    mensaje = "Respuesta vacía del servidor"
                )
            } else {
                // Intentar parsear el error del body
                val errorBody = response.errorBody()?.string()
                if (errorBody != null) {
                    try {
                        Gson().fromJson(errorBody, LoginResponseDTO::class.java)
                    } catch (e: Exception) {
                        LoginResponseDTO(
                            exitoso = false,
                            mensaje = "Error HTTP ${response.code()}"
                        )
                    }
                } else {
                    LoginResponseDTO(
                        exitoso = false,
                        mensaje = "Error HTTP ${response.code()}"
                    )
                }
            }
        } catch (e: Exception) {
            LoginResponseDTO(
                exitoso = false,
                mensaje = "Error de conexión: ${e.message}"
            )
        }
    }
}
